classad/_expression.py

```diff
@@ -0,0 +1,299 @@
+import operator
+from collections import MutableMapping, OrderedDict
+
+import pyparsing as pp
+from typing import Any, Iterable, List, Iterator
+
+from classad._operator import eq_operator, ne_operator, not_operator
+from classad._primitives import Error, Undefined, HTCBool
+from ._base_expression import Expression
+from . import _functions
+
+
+def scope_up(key: List[str]):
+    return key[:-1]
+
+
+class ClassAd(Expression, MutableMapping):
+    __slots__ = "_data"
+
+    def __add__(self, other):
+        return Error()
+
+    __sub__ = __rsub__ = __radd__ = __add__
+
+    def __init__(self):
+        super().__init__()
+        self._data = OrderedDict()
```

Comment:
The standard ``dict`` is de-facto ordered on Python 3.6 (CPython since 3.6, PyPy since 2.7) and faster. I recommend we use it instead.

```suggestion
        self._data = dict()
```